pkg/ccl/crosscluster/logical/lww_row_processor_test.go

```diff
@@ -301,3 +306,165 @@ func BenchmarkLWWInsertBatch(b *testing.B) {
+	tableNumber := 0
+	createTable := func(t *testing.T) string {
+		tableName := fmt.Sprintf("tab%d", tableNumber)
+		runner.Exec(t, fmt.Sprintf(`CREATE TABLE %s (pk int primary key, payload string)`, tableName))
+		runner.Exec(t, fmt.Sprintf(
+			"ALTER TABLE %s "+lwwColumnAdd,
+			tableName))
+		tableNumber++
+		return tableName
+	}
+
+	setup := func(t *testing.T, useKVProc bool) (string, RowProcessor, func(hlc.Timestamp, ...interface{}) roachpb.KeyValue) {
+		tableNameSrc := createTable(t)
+		tableNameDst := createTable(t)
+		srcDesc := desctestutils.TestingGetPublicTableDescriptor(s.DB(), s.Codec(), "defaultdb", tableNameSrc)
+		dstDesc := desctestutils.TestingGetPublicTableDescriptor(s.DB(), s.Codec(), "defaultdb", tableNameDst)
+
+		// We need the SQL row processor even when testing the KW row processor since it's the fallback
+		var rp RowProcessor
+		rp, err := makeSQLProcessor(ctx, s.ClusterSettings(), map[descpb.ID]sqlProcessorTableConfig{
+			dstDesc.GetID(): {
+				srcDesc: srcDesc,
+			},
+		}, jobspb.JobID(1), s.InternalExecutor().(isql.Executor))
+		require.NoError(t, err)
+
+		if useKVProc {
+			rp, err = newKVRowProcessor(ctx,
+				&execinfra.ServerConfig{
+					DB:           s.InternalDB().(descs.DB),
+					LeaseManager: s.LeaseManager(),
+				}, &eval.Context{
+					Codec:    s.Codec(),
+					Settings: s.ClusterSettings(),
+				}, map[descpb.ID]sqlProcessorTableConfig{
+					dstDesc.GetID(): {
+						srcDesc: srcDesc,
+					},
+				},
+				rp.(*sqlRowProcessor))
+			require.NoError(t, err)
+		}
+		return tableNameDst, rp, func(timestamp hlc.Timestamp, datums ...interface{}) roachpb.KeyValue {
+			kv := replicationtestutils.EncodeKV(t, s.Codec(), srcDesc, datums...)
+			kv.Value.Timestamp = timestamp
+			return kv
+		}
+	}
+
+	insertRow := func(rp RowProcessor, keyValue roachpb.KeyValue, prevValue roachpb.Value) error {
+		return s.InternalDB().(isql.DB).Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+			_, err := rp.ProcessRow(ctx, txn, keyValue, prevValue)
+			return err
+		})
+	}
+
+	timeNow := hlc.Timestamp{WallTime: timeutil.Now().UnixNano()}
+	timeOneDayForward := hlc.Timestamp{WallTime: timeutil.Now().Add(time.Hour * 24).UnixNano()}
+	row1 := []interface{}{1, "row1"}
+	row2 := []interface{}{1, "row2"}
+	row3 := []interface{}{1, "row3"}
+
+	testutils.RunTrueAndFalse(t, "useKVProc", func(t *testing.T, useKVProc bool) {
+		testutils.RunTrueAndFalse(t, "optimistic_insert", func(t *testing.T, optimisticInsert bool) {
+
+			t.Run("cross-cluster-write", func(t *testing.T) {
+				runner.Exec(t, fmt.Sprintf("SET CLUSTER SETTING logical_replication.consumer.try_optimistic_insert.enabled=%t", optimisticInsert))
+				defer runner.Exec(t, "RESET CLUSTER SETTING logical_replication.consumer.try_optimistic_insert.enabled")
+				tableNameDst, rp, encoder := setup(t, useKVProc)
+
+				runner.Exec(t, fmt.Sprintf("INSERT INTO %s VALUES ($1, $2)", tableNameDst), row1...)
+
+				keyValue2 := encoder(timeOneDayForward, row2...)
```

Comment:
before inserting row2 with `timeOneDayForward`, could you try inserting row2 with `timeOneDayBackward` and assert that row 1 wins?

Reply:
Done